Cython/Compiler/ModuleNode.py

```diff
@@ -1865,15 +1865,19 @@ def generate_dealloc_function(self, scope, code):
                 code.putln("} else")
                 code.putln("#endif")
                 code.putln("{")
+            code.putln("PyTypeObject *tp = Py_TYPE(o);")
             code.putln("#if CYTHON_USE_TYPE_SLOTS")
             # Asking for PyType_GetSlot(..., Py_tp_free) seems to cause an error in pypy
-            code.putln("(*Py_TYPE(o)->tp_free)(o);")
+            code.putln("(*tp->tp_free)(o);")
             code.putln("#else")
             code.putln("{")
-            code.putln("freefunc tp_free = (freefunc)PyType_GetSlot(Py_TYPE(o), Py_tp_free);")
+            code.putln("freefunc tp_free = (freefunc)PyType_GetSlot(tp, Py_tp_free);")
             code.putln("if (tp_free) tp_free(o);")
             code.putln("}")
             code.putln("#endif")
+            code.putln("#if CYTHON_USE_TYPE_SPECS")
+            code.putln("Py_DECREF(tp);")
+            code.putln("#endif")
```

Comment:
When I saw this, I thought: what, you take a borrowed reference above and then decref it - how's that a fix? And it took me a while to remember that because we're increfing the type reference in `tp_new()`, you're not intending the decref `tp` here but really `o->ob_type` (which is technically the same but semantically different). I think that's worth a code comment.

Also, don't we need to do the same thing for objects that go to the freelist? Your change only goes into the `else` branch.

Reply:
> Also, don't we need to do the same thing for objects that go to the freelist? Your change only goes into the else branch.

Yes we should.

We're more restrictive than we need to be on freelists for heaptypes so it's surprisingly hard to get one. I'll do a separate PR to relax that.